'미국의 소득과 종교' 데이터,

종교(religion) / 나머지 변수

In [48]:
import pandas as pd
pew = pd.read_csv("C:/Users/user/Documents/data/pew.csv")
pew.head()

religion  <$10k  $10-20k  $20-30k  $30-40k  $40-50k  $50-75k  \
0            Agnostic     27       34       60       81       76      137   
1             Atheist     12       27       37       52       35       70   
2            Buddhist     27       21       30       34       33       58   
3            Catholic    418      617      732      670      638     1116   
4  Don’t know/refused     15       14       15       11       10       35   

   $75-100k  $100-150k  >150k  Don't know/refused  
0       122        109     84                  96  
1        73         59     74                  76  
2        62         39     53                  54  
3       949        792    633                1489  
4        21         17     18                 116

6개 열만 출력

In [49]:
pew.iloc[:,0:6].head()

religion  <$10k  $10-20k  $20-30k  $30-40k  $40-50k
0            Agnostic     27       34       60       81       76
1             Atheist     12       27       37       52       35
2            Buddhist     27       21       30       34       33
3            Catholic    418      617      732      670      638
4  Don’t know/refused     15       14       15       11       10

melt 메서드의 인자

    id vars : 위치를 그대로 유지할 열의 이름을 지정
    value_vars : 행으로 위치를 변경할 열의 이름을 지정
    var_name : value_vars로 위치를 변경한 열의 이름을 지정
    value_name : var_name으로 위치를 변경한 열의 데이터를 저장한 열의 이름을 지정

id_vars 인잣값으로 지정한 열('religion')을 제외한 나머지 소득 정보 열('10k~150k')이 'variable(변수)' 열로 정리되고 소득정보 열의 행도 데이터도 value 열로 정리됌.

== "reglion열을 고정하여 피벗했다"

    pew_long = pd.melt(pew, id_vars='religion')
    pew_long

In [50]:
pew_long = pd.melt(pew, id_vars='religion')
pew_long

religion            variable  value
0                 Agnostic               <$10k     27
1                  Atheist               <$10k     12
2                 Buddhist               <$10k     27
3                 Catholic               <$10k    418
4       Don’t know/refused               <$10k     15
..                     ...                 ...    ...
175               Orthodox  Don't know/refused     73
176        Other Christian  Don't know/refused     18
177           Other Faiths  Don't know/refused     71
178  Other World Religions  Don't know/refused      8
179           Unaffiliated  Don't know/refused    597

[180 rows x 3 columns]

varaible, value 명을 income, count로 바꾸기

In [51]:
pew_long = pd.melt(pew, id_vars='religion', var_name='income', value_name='count')
pew_long

religion              income  count
0                 Agnostic               <$10k     27
1                  Atheist               <$10k     12
2                 Buddhist               <$10k     27
3                 Catholic               <$10k    418
4       Don’t know/refused               <$10k     15
..                     ...                 ...    ...
175               Orthodox  Don't know/refused     73
176        Other Christian  Don't know/refused     18
177           Other Faiths  Don't know/refused     71
178  Other World Religions  Don't know/refused      8
179           Unaffiliated  Don't know/refused    597

[180 rows x 3 columns]

In [52]:
print( pew_long[pew_long['income']=='< $10k'].value_counts().head() )
print('\n')
print( pew_long[pew_long['income']=="Don't know/refused"].value_counts() )

Series([], dtype: int64)


religion                 income              count
Agnostic                 Don't know/refused  96       1
Atheist                  Don't know/refused  76       1
Other World Religions    Don't know/refused  8        1
Other Faiths             Don't know/refused  71       1
Other Christian          Don't know/refused  18       1
Orthodox                 Don't know/refused  73       1
Muslim                   Don't know/refused  22       1
Mormon                   Don't know/refused  69       1
Mainline Prot            Don't know/refused  1328     1
Jewish                   Don't know/refused  162      1
Jehovah's Witness        Don't know/refused  37       1
Historically Black Prot  Don't know/refused  339      1
Hindu                    Don't know/refused  37       1
Evangelical Prot         Don't know/refused  1529     1
Don’t know/refused       Don't know/refused  116      1
Catholic                 Don't know/refused  1489     1
Buddhist                 D

2개 이상의 열을 고정하고 나머지 열을 행으로 바꾸기

In [53]:
import pandas as pd
path = "C:/Users/user/Documents/data"
billboard = pd.read_csv(path + "/billboard.csv")
billboard.head()

year        artist                    track  time date.entered  wk1   wk2  \
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26   87  82.0   
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02   91  87.0   
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08   81  70.0   
3  2000  3 Doors Down                    Loser  4:24   2000-10-21   76  76.0   
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15   57  34.0   

    wk3   wk4   wk5  ...  wk67  wk68  wk69  wk70  wk71  wk72  wk73  wk74  \
0  72.0  77.0  87.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1  92.0   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2  68.0  67.0  66.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
3  72.0  69.0  67.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  25.0  17.0  17.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   wk75  wk76  
0   NaN   NaN  
1   NaN   NaN  
2   NaN   NaN  
3   NaN   NaN  
4   NaN   NaN  

[5 rows x 81 columns]

year	artist	track	time	date.entered 열을 모두 고정하고 나머지 열(wk1	wk2	wk3	wk4	wk5	...	wk67	wk68	wk69	wk70	wk71	wk72	wk73	wk74	wk75	wk76) 피벗하기 
        
       * 아마 wk1~76이 week 에 대한 rating. 

In [54]:
pd.melt(billboard, id_vars=['year','artist','track','time','date.entered'])
pd.melt(billboard, id_vars=['year','artist','track','time','date.entered'], var_name='week', value_name='rating')

year            artist                    track  time date.entered  \
0      2000             2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26   
1      2000           2Ge+her  The Hardest Part Of ...  3:15   2000-09-02   
2      2000      3 Doors Down               Kryptonite  3:53   2000-04-08   
3      2000      3 Doors Down                    Loser  4:24   2000-10-21   
4      2000          504 Boyz            Wobble Wobble  3:35   2000-04-15   
...     ...               ...                      ...   ...          ...   
24087  2000       Yankee Grey     Another Nine Minutes  3:10   2000-04-29   
24088  2000  Yearwood, Trisha          Real Live Woman  3:55   2000-04-01   
24089  2000   Ying Yang Twins  Whistle While You Tw...  4:19   2000-03-18   
24090  2000     Zombie Nation            Kernkraft 400  3:30   2000-09-02   
24091  2000   matchbox twenty                     Bent  4:12   2000-04-29   

       week  rating  
0       wk1    87.0  
1       wk1    91.0  
2       wk1    81.0  
3       wk1    76.0  
4       wk1    57.0  
...     ...     ...  
24087  wk76     NaN  
24088  wk76     NaN  
24089  wk76     NaN  
24090  wk76     NaN  
24091  wk76     NaN  

[24092 rows x 7 columns]

Deaths_Guinea는 '사망자 수'와 '나라 이름' 합쳐 만든거

In [55]:
ebola = pd.read_csv(path + "/country_timeseries.csv")
ebola.head()

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
0    1/5/2015  289        2776.0            NaN            10030.0   
1    1/4/2015  288        2775.0            NaN             9780.0   
2    1/3/2015  287        2769.0         8166.0             9722.0   
3    1/2/2015  286           NaN         8157.0                NaN   
4  12/31/2014  284        2730.0         8115.0             9633.0   

   Cases_Nigeria  Cases_Senegal  Cases_UnitedStates  Cases_Spain  Cases_Mali  \
0            NaN            NaN                 NaN          NaN         NaN   
1            NaN            NaN                 NaN          NaN         NaN   
2            NaN            NaN                 NaN          NaN         NaN   
3            NaN            NaN                 NaN          NaN         NaN   
4            NaN            NaN                 NaN          NaN         NaN   

   Deaths_Guinea  Deaths_Liberia  Deaths_SierraLeone  Deaths_Nigeria  \
0         1786.0             NaN              2977.0             NaN   
1         1781.0             NaN              2943.0             NaN   
2         1767.0          3496.0              2915.0             NaN   
3            NaN          3496.0                 NaN             NaN   
4         1739.0          3471.0              2827.0             NaN   

   Deaths_Senegal  Deaths_UnitedStates  Deaths_Spain  Deaths_Mali  
0             NaN                  NaN           NaN          NaN  
1             NaN                  NaN           NaN          NaN  
2             NaN                  NaN           NaN          NaN  
3             NaN                  NaN           NaN          NaN  
4             NaN                  NaN           NaN          NaN

0, 1, 2, 3, 10, 11 열의 5개 데이터만 확인해보기

In [56]:
ebola.iloc[0 : 5 , [0, 1, 2, 3, 10, 11]]

Date  Day  Cases_Guinea  Cases_Liberia  Deaths_Guinea  Deaths_Liberia
0    1/5/2015  289        2776.0            NaN         1786.0             NaN
1    1/4/2015  288        2775.0            NaN         1781.0             NaN
2    1/3/2015  287        2769.0         8166.0         1767.0          3496.0
3    1/2/2015  286           NaN         8157.0            NaN          3496.0
4  12/31/2014  284        2730.0         8115.0         1739.0          3471.0

Date 와 Day를 고정하고 나머지 행으로 피벗하기 (Date, Day를 제외한 나머지 컬럼이 어차피 Cases_국가명 , Deathes_국가명의 나열 이기 때문)

In [57]:
ebola_long = pd.melt(ebola, id_vars=['Date','Day'])
ebola_long 

Date  Day      variable   value
0       1/5/2015  289  Cases_Guinea  2776.0
1       1/4/2015  288  Cases_Guinea  2775.0
2       1/3/2015  287  Cases_Guinea  2769.0
3       1/2/2015  286  Cases_Guinea     NaN
4     12/31/2014  284  Cases_Guinea  2730.0
...          ...  ...           ...     ...
1947   3/27/2014    5   Deaths_Mali     NaN
1948   3/26/2014    4   Deaths_Mali     NaN
1949   3/25/2014    3   Deaths_Mali     NaN
1950   3/24/2014    2   Deaths_Mali     NaN
1951   3/22/2014    0   Deaths_Mali     NaN

[1952 rows x 4 columns]

In [58]:
ebola_long.variable

0       Cases_Guinea
1       Cases_Guinea
2       Cases_Guinea
3       Cases_Guinea
4       Cases_Guinea
            ...     
1947     Deaths_Mali
1948     Deaths_Mali
1949     Deaths_Mali
1950     Deaths_Mali
1951     Deaths_Mali
Name: variable, Length: 1952, dtype: object

In [59]:
ebola_long.variable.split('_')

AttributeError: 'Series' object has no attribute 'split'

In [60]:
ebola_long.variable.split('_').str

AttributeError: 'Series' object has no attribute 'split'

split 메서드로 열 이름 분리

In [61]:
variable_split = ebola_long.variable.str.split('_')

print(variable_split.head())
print('\n')
print(variable_split[ : 5])

0    [Cases, Guinea]
1    [Cases, Guinea]
2    [Cases, Guinea]
3    [Cases, Guinea]
4    [Cases, Guinea]
Name: variable, dtype: object


0    [Cases, Guinea]
1    [Cases, Guinea]
2    [Cases, Guinea]
3    [Cases, Guinea]
4    [Cases, Guinea]
Name: variable, dtype: object


In [62]:
print(variable_split[ : 5])
print('\n')
print(type(variable_split))
print('\n')
print(variable_split[3])
print('\n')
print(type(variable_split[0]))

0    [Cases, Guinea]
1    [Cases, Guinea]
2    [Cases, Guinea]
3    [Cases, Guinea]
4    [Cases, Guinea]
Name: variable, dtype: object


<class 'pandas.core.series.Series'>


['Cases', 'Guinea']


<class 'list'>


발병/죽음, 국가

In [63]:
print(variable_split[ : 5])

0    [Cases, Guinea]
1    [Cases, Guinea]
2    [Cases, Guinea]
3    [Cases, Guinea]
4    [Cases, Guinea]
Name: variable, dtype: object


In [64]:
status_values = variable_split.str.get(0)
status_values

print(status_values.iloc[:5])
print('\n')
print(status_values.iloc[-5:])

0    Cases
1    Cases
2    Cases
3    Cases
4    Cases
Name: variable, dtype: object


1947    Deaths
1948    Deaths
1949    Deaths
1950    Deaths
1951    Deaths
Name: variable, dtype: object


In [65]:
country_values = variable_split.str.get(1)
country_values

print(country_values.iloc[:5])
print('\n')
print(country_values.iloc[-5:])

0    Guinea
1    Guinea
2    Guinea
3    Guinea
4    Guinea
Name: variable, dtype: object


1947    Mali
1948    Mali
1949    Mali
1950    Mali
1951    Mali
Name: variable, dtype: object


In [66]:
ebola_long['status'] = status_values
ebola_long['country'] = country_values
ebola_long.head()

Date  Day      variable   value status country
0    1/5/2015  289  Cases_Guinea  2776.0  Cases  Guinea
1    1/4/2015  288  Cases_Guinea  2775.0  Cases  Guinea
2    1/3/2015  287  Cases_Guinea  2769.0  Cases  Guinea
3    1/2/2015  286  Cases_Guinea     NaN  Cases  Guinea
4  12/31/2014  284  Cases_Guinea  2730.0  Cases  Guinea

concat 메서드로 데이터프레임에 열 추가하기

In [67]:
ebola_long.head()

Date  Day      variable   value status country
0    1/5/2015  289  Cases_Guinea  2776.0  Cases  Guinea
1    1/4/2015  288  Cases_Guinea  2775.0  Cases  Guinea
2    1/3/2015  287  Cases_Guinea  2769.0  Cases  Guinea
3    1/2/2015  286  Cases_Guinea     NaN  Cases  Guinea
4  12/31/2014  284  Cases_Guinea  2730.0  Cases  Guinea

In [68]:
variable_split = ebola_long.variable.str.split('_', expand = True)
variable_split

0       1
0      Cases  Guinea
1      Cases  Guinea
2      Cases  Guinea
3      Cases  Guinea
4      Cases  Guinea
...      ...     ...
1947  Deaths    Mali
1948  Deaths    Mali
1949  Deaths    Mali
1950  Deaths    Mali
1951  Deaths    Mali

[1952 rows x 2 columns]

In [69]:
variable_split.columns = ['status','country']
variable_split

status country
0      Cases  Guinea
1      Cases  Guinea
2      Cases  Guinea
3      Cases  Guinea
4      Cases  Guinea
...      ...     ...
1947  Deaths    Mali
1948  Deaths    Mali
1949  Deaths    Mali
1950  Deaths    Mali
1951  Deaths    Mali

[1952 rows x 2 columns]

In [70]:
ebola_parsed = pd.concat([ebola_long, variable_split], axis = 1) # axis = 1 : 열
ebola_parsed

Date  Day      variable   value  status country  status country
0       1/5/2015  289  Cases_Guinea  2776.0   Cases  Guinea   Cases  Guinea
1       1/4/2015  288  Cases_Guinea  2775.0   Cases  Guinea   Cases  Guinea
2       1/3/2015  287  Cases_Guinea  2769.0   Cases  Guinea   Cases  Guinea
3       1/2/2015  286  Cases_Guinea     NaN   Cases  Guinea   Cases  Guinea
4     12/31/2014  284  Cases_Guinea  2730.0   Cases  Guinea   Cases  Guinea
...          ...  ...           ...     ...     ...     ...     ...     ...
1947   3/27/2014    5   Deaths_Mali     NaN  Deaths    Mali  Deaths    Mali
1948   3/26/2014    4   Deaths_Mali     NaN  Deaths    Mali  Deaths    Mali
1949   3/25/2014    3   Deaths_Mali     NaN  Deaths    Mali  Deaths    Mali
1950   3/24/2014    2   Deaths_Mali     NaN  Deaths    Mali  Deaths    Mali
1951   3/22/2014    0   Deaths_Mali     NaN  Deaths    Mali  Deaths    Mali

[1952 rows x 8 columns]

In [71]:
weather  = pd.read_csv(path + "/weather.csv")
weather.iloc[ :5, : ]

id  year  month element  d1    d2    d3  d4    d5  d6  ...  d22   d23  \
0  MX17004  2010      1    tmax NaN   NaN   NaN NaN   NaN NaN  ...  NaN   NaN   
1  MX17004  2010      1    tmin NaN   NaN   NaN NaN   NaN NaN  ...  NaN   NaN   
2  MX17004  2010      2    tmax NaN  27.3  24.1 NaN   NaN NaN  ...  NaN  29.9   
3  MX17004  2010      2    tmin NaN  14.4  14.4 NaN   NaN NaN  ...  NaN  10.7   
4  MX17004  2010      3    tmax NaN   NaN   NaN NaN  32.1 NaN  ...  NaN   NaN   

   d24  d25  d26  d27  d28  d29   d30  d31  
0  NaN  NaN  NaN  NaN  NaN  NaN  27.8  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  14.5  NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN   NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN   NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN   NaN  NaN  

[5 rows x 35 columns]

melt 메서드로 일별 온도 측정값(d1,d2, ...)을 피벗하시오.

In [72]:
weather_melt = pd.melt(weather, id_vars=['id','year','month','element'], var_name='day', value_name='temp')
print(weather_melt.head())

        id  year  month element day  temp
0  MX17004  2010      1    tmax  d1   NaN
1  MX17004  2010      1    tmin  d1   NaN
2  MX17004  2010      2    tmax  d1   NaN
3  MX17004  2010      2    tmin  d1   NaN
4  MX17004  2010      3    tmax  d1   NaN


"pivot_table 메서드는 행과 열의 위치를 다시 바꿔 정리해준다. index 인자에는 위치를 그대로 유지할 열 이름을 지정하고, 
columns 인자에는 피벗할 열 이름을 지정하고, values 인자에는 새로운 열의 데이터가 될 열 이름을 지정하면 된다."

In [73]:
weather_tidy = weather_melt.pivot_table(index=['id','year','month','element'])
weather_tidy

temp
id      year month element           
MX17004 2010 1     tmax     27.800000
                   tmin     14.500000
             2     tmax     27.750000
                   tmin     13.225000
             3     tmax     32.566667
                   tmin     16.200000
             4     tmax     36.300000
                   tmin     16.700000
             5     tmax     33.200000
                   tmin     18.200000
             6     tmax     29.050000
                   tmin     17.750000
             7     tmax     29.250000
                   tmin     17.000000
             8     tmax     28.271429
                   tmin     15.842857
             10    tmax     28.900000
                   tmin     13.080000
             11    tmax     28.120000
                   tmin     12.500000
             12    tmax     28.850000
                   tmin     12.150000

In [74]:
weather_tidy = weather_melt.pivot_table(index=['id','year','month','element'], columns='element')
print ( weather_tidy )

                                 temp           
element                          tmax       tmin
id      year month element                      
MX17004 2010 1     tmax     27.800000        NaN
                   tmin           NaN  14.500000
             2     tmax     27.750000        NaN
                   tmin           NaN  13.225000
             3     tmax     32.566667        NaN
                   tmin           NaN  16.200000
             4     tmax     36.300000        NaN
                   tmin           NaN  16.700000
             5     tmax     33.200000        NaN
                   tmin           NaN  18.200000
             6     tmax     29.050000        NaN
                   tmin           NaN  17.750000
             7     tmax     29.250000        NaN
                   tmin           NaN  17.000000
             8     tmax     28.271429        NaN
                   tmin           NaN  15.842857
             10    tmax     28.900000        NaN
                   t

In [75]:
weather_tidy = weather_melt.pivot_table(index=['id','year','month','element'], columns='element', values='temp') #
print (weather_tidy )

element                          tmax       tmin
id      year month element                      
MX17004 2010 1     tmax     27.800000        NaN
                   tmin           NaN  14.500000
             2     tmax     27.750000        NaN
                   tmin           NaN  13.225000
             3     tmax     32.566667        NaN
                   tmin           NaN  16.200000
             4     tmax     36.300000        NaN
                   tmin           NaN  16.700000
             5     tmax     33.200000        NaN
                   tmin           NaN  18.200000
             6     tmax     29.050000        NaN
                   tmin           NaN  17.750000
             7     tmax     29.250000        NaN
                   tmin           NaN  17.000000
             8     tmax     28.271429        NaN
                   tmin           NaN  15.842857
             10    tmax     28.900000        NaN
                   tmin           NaN  13.080000
             11    t

In [76]:
weather_tidy_flat = weather_tidy.reset_index()    # reset_index()
print( weather_tidy_flat )

element       id  year  month element       tmax       tmin
0        MX17004  2010      1    tmax  27.800000        NaN
1        MX17004  2010      1    tmin        NaN  14.500000
2        MX17004  2010      2    tmax  27.750000        NaN
3        MX17004  2010      2    tmin        NaN  13.225000
4        MX17004  2010      3    tmax  32.566667        NaN
5        MX17004  2010      3    tmin        NaN  16.200000
6        MX17004  2010      4    tmax  36.300000        NaN
7        MX17004  2010      4    tmin        NaN  16.700000
8        MX17004  2010      5    tmax  33.200000        NaN
9        MX17004  2010      5    tmin        NaN  18.200000
10       MX17004  2010      6    tmax  29.050000        NaN
11       MX17004  2010      6    tmin        NaN  17.750000
12       MX17004  2010      7    tmax  29.250000        NaN
13       MX17004  2010      7    tmin        NaN  17.000000
14       MX17004  2010      8    tmax  28.271429        NaN
15       MX17004  2010      8    tmin   

In [77]:
import pandas as pd
path = "C:/Users/user/Documents/data"
billboard = pd.read_csv(path + "/billboard.csv")
billboard.head()

year        artist                    track  time date.entered  wk1   wk2  \
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26   87  82.0   
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02   91  87.0   
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08   81  70.0   
3  2000  3 Doors Down                    Loser  4:24   2000-10-21   76  76.0   
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15   57  34.0   

    wk3   wk4   wk5  ...  wk67  wk68  wk69  wk70  wk71  wk72  wk73  wk74  \
0  72.0  77.0  87.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
1  92.0   NaN   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2  68.0  67.0  66.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
3  72.0  69.0  67.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  25.0  17.0  17.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

   wk75  wk76  
0   NaN   NaN  
1   NaN   NaN  
2   NaN   NaN  
3   NaN   NaN  
4   NaN   NaN  

[5 rows x 81 columns]

In [78]:
billboard_long = pd.melt(billboard, id_vars=['year','artist','track','time','date.entered'], var_name='week', value_name='rating')
billboard_long

year            artist                    track  time date.entered  \
0      2000             2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26   
1      2000           2Ge+her  The Hardest Part Of ...  3:15   2000-09-02   
2      2000      3 Doors Down               Kryptonite  3:53   2000-04-08   
3      2000      3 Doors Down                    Loser  4:24   2000-10-21   
4      2000          504 Boyz            Wobble Wobble  3:35   2000-04-15   
...     ...               ...                      ...   ...          ...   
24087  2000       Yankee Grey     Another Nine Minutes  3:10   2000-04-29   
24088  2000  Yearwood, Trisha          Real Live Woman  3:55   2000-04-01   
24089  2000   Ying Yang Twins  Whistle While You Tw...  4:19   2000-03-18   
24090  2000     Zombie Nation            Kernkraft 400  3:30   2000-09-02   
24091  2000   matchbox twenty                     Bent  4:12   2000-04-29   

       week  rating  
0       wk1    87.0  
1       wk1    91.0  
2       wk1    81.0  
3       wk1    76.0  
4       wk1    57.0  
...     ...     ...  
24087  wk76     NaN  
24088  wk76     NaN  
24089  wk76     NaN  
24090  wk76     NaN  
24091  wk76     NaN  

[24092 rows x 7 columns]

In [79]:
print(billboard_long[billboard_long.track == 'Loser'].head())
print(billboard_long[billboard_long.track == 'Real Live Woman'].head())
print(billboard_long[billboard_long.track == 'Bent'].head())
print(billboard_long.shape)

      year        artist  track  time date.entered week  rating
3     2000  3 Doors Down  Loser  4:24   2000-10-21  wk1    76.0
320   2000  3 Doors Down  Loser  4:24   2000-10-21  wk2    76.0
637   2000  3 Doors Down  Loser  4:24   2000-10-21  wk3    72.0
954   2000  3 Doors Down  Loser  4:24   2000-10-21  wk4    69.0
1271  2000  3 Doors Down  Loser  4:24   2000-10-21  wk5    67.0
      year            artist            track  time date.entered week  rating
313   2000  Yearwood, Trisha  Real Live Woman  3:55   2000-04-01  wk1    85.0
630   2000  Yearwood, Trisha  Real Live Woman  3:55   2000-04-01  wk2    83.0
947   2000  Yearwood, Trisha  Real Live Woman  3:55   2000-04-01  wk3    83.0
1264  2000  Yearwood, Trisha  Real Live Woman  3:55   2000-04-01  wk4    82.0
1581  2000  Yearwood, Trisha  Real Live Woman  3:55   2000-04-01  wk5    81.0
      year           artist track  time date.entered week  rating
316   2000  matchbox twenty  Bent  4:12   2000-04-29  wk1    60.0
633   2000  matc

  year        artist  track  time date.entered  중복제거

In [80]:
billboard_songs = billboard_long[['year','artist','track','time','date.entered']]
print(billboard_songs.shape)

(24092, 5)


중복제거 함수 .drop_duplicates()

In [81]:
billboard_songs = billboard_songs.drop_duplicates() 
print( billboard_songs )

     year            artist                    track  time date.entered
0    2000             2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26
1    2000           2Ge+her  The Hardest Part Of ...  3:15   2000-09-02
2    2000      3 Doors Down               Kryptonite  3:53   2000-04-08
3    2000      3 Doors Down                    Loser  4:24   2000-10-21
4    2000          504 Boyz            Wobble Wobble  3:35   2000-04-15
..    ...               ...                      ...   ...          ...
312  2000       Yankee Grey     Another Nine Minutes  3:10   2000-04-29
313  2000  Yearwood, Trisha          Real Live Woman  3:55   2000-04-01
314  2000   Ying Yang Twins  Whistle While You Tw...  4:19   2000-03-18
315  2000     Zombie Nation            Kernkraft 400  3:30   2000-09-02
316  2000   matchbox twenty                     Bent  4:12   2000-04-29

[317 rows x 5 columns]


In [82]:
billboard_songs['id'] = range(len(billboard_songs))
billboard_songs.iloc[:10]

year          artist                    track  time date.entered  id
0  2000           2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26   0
1  2000         2Ge+her  The Hardest Part Of ...  3:15   2000-09-02   1
2  2000    3 Doors Down               Kryptonite  3:53   2000-04-08   2
3  2000    3 Doors Down                    Loser  4:24   2000-10-21   3
4  2000        504 Boyz            Wobble Wobble  3:35   2000-04-15   4
5  2000            98^0  Give Me Just One Nig...  3:24   2000-08-19   5
6  2000         A*Teens            Dancing Queen  3:44   2000-07-08   6
7  2000         Aaliyah            I Don't Wanna  4:15   2000-01-29   7
8  2000         Aaliyah                Try Again  4:03   2000-03-18   8
9  2000  Adams, Yolanda            Open My Heart  5:30   2000-08-26   9

In [83]:
len(billboard_songs)

317

In [84]:
range(len(billboard_songs))

range(0, 317)

In [93]:
 billboard_long.iloc[:10]

year          artist                    track  time date.entered week  \
0  2000           2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1   
1  2000         2Ge+her  The Hardest Part Of ...  3:15   2000-09-02  wk1   
2  2000    3 Doors Down               Kryptonite  3:53   2000-04-08  wk1   
3  2000    3 Doors Down                    Loser  4:24   2000-10-21  wk1   
4  2000        504 Boyz            Wobble Wobble  3:35   2000-04-15  wk1   
5  2000            98^0  Give Me Just One Nig...  3:24   2000-08-19  wk1   
6  2000         A*Teens            Dancing Queen  3:44   2000-07-08  wk1   
7  2000         Aaliyah            I Don't Wanna  4:15   2000-01-29  wk1   
8  2000         Aaliyah                Try Again  4:03   2000-03-18  wk1   
9  2000  Adams, Yolanda            Open My Heart  5:30   2000-08-26  wk1   

   rating  
0    87.0  
1    91.0  
2    81.0  
3    76.0  
4    57.0  
5    51.0  
6    97.0  
7    84.0  
8    59.0  
9    76.0

merge 개념이 헷갈림. (메모용)

In [94]:
billboard_ratings = billboard_long.merge(billboard_songs, on = ['year', 'artist', 'track', 'time'])
billboard_ratings.head(10)

year artist                    track  time date.entered_x  week  rating  \
0  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk1    87.0   
1  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk2    82.0   
2  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk3    72.0   
3  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk4    77.0   
4  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk5    87.0   
5  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk6    94.0   
6  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk7    99.0   
7  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk8     NaN   
8  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26   wk9     NaN   
9  2000  2 Pac  Baby Don't Cry (Keep...  4:22     2000-02-26  wk10     NaN   

  date.entered_y  id  
0     2000-02-26   0  
1     2000-02-26   0  
2     2000-02-26   0  
3     2000-02-26   0  
4     2000-02-26   0  
5     2000-02-26   0  
6     2000-02-26   0  
7     2000-02-26   0  
8     2000-02-26   0  
9     2000-02-26   0

대용량 데이터 처리하기

enumerate 헷갈림 (메모용)

아래 코드 전반적으로 햇갈림

In [96]:
import os
import urllib.request

path = 'C:/Users/user/Documents/data'

with open(path + '/raw_data_urls.txt', 'r') as data_urls:
    for line, url in enumerate(data_urls):
        if line == 5:
            break
        fn = url.split('/')[-1].strip()
        fp = os.path.join('', '../data', fn)
        print(url)
        print(fp)
        urllib.request.urlretrieve(url, fp)

https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-01.csv

../data\fhv_tripdata_2015-01.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-02.csv

../data\fhv_tripdata_2015-02.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-03.csv

../data\fhv_tripdata_2015-03.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-04.csv

../data\fhv_tripdata_2015-04.csv
https://s3.amazonaws.com/nyc-tlc/trip+data/fhv_tripdata_2015-05.csv

../data\fhv_tripdata_2015-05.csv


glot 라이브러리에 포함된 glob 메서드는 특정한 패턴의 이름을 가진 파일을 한 번에 읽어 들일 수 있다. 

In [98]:
import glob
nyc_taxi_data = glob.glob('../data/fhv_*')
print(nyc_taxi_data)

['../data\\fhv_tripdata_2015-01.csv', '../data\\fhv_tripdata_2015-02.csv', '../data\\fhv_tripdata_2015-03.csv', '../data\\fhv_tripdata_2015-04.csv', '../data\\fhv_tripdata_2015-05.csv']


In [99]:
taxi1 = pd.read_csv(nyc_taxi_data[0])
taxi2 = pd.read_csv(nyc_taxi_data[1])
taxi3 = pd.read_csv(nyc_taxi_data[2])
taxi4 = pd.read_csv(nyc_taxi_data[3])
taxi5 = pd.read_csv(nyc_taxi_data[4])

In [103]:
print(taxi1.head(2))
print('\n')
print(taxi2.head(2))
print('\n')
print(taxi3.head(2))
print('\n')
print(taxi4.head(2))
print('\n')
print(taxi5.head(2))
print('\n')

print(taxi1.shape)
print('\n')
print(taxi2.shape)
print('\n')
print(taxi3.shape)
print('\n')
print(taxi4.shape)
print('\n')
print(taxi5.shape)

  Dispatching_base_num          Pickup_date  locationID
0               B00013  2015-01-01 00:30:00         NaN
1               B00013  2015-01-01 01:22:00         NaN


  Dispatching_base_num          Pickup_date  locationID
0               B00013  2015-02-01 00:00:00         NaN
1               B00013  2015-02-01 00:01:00         NaN


  Dispatching_base_num          Pickup_date  locationID
0               B00029  2015-03-01 00:02:00       213.0
1               B00029  2015-03-01 00:03:00        51.0


  Dispatching_base_num          Pickup_date  locationID
0               B00001  2015-04-01 04:30:00         NaN
1               B00001  2015-04-01 06:00:00         NaN


  Dispatching_base_num          Pickup_date  locationID
0               B00001  2015-05-01 04:30:00         NaN
1               B00001  2015-05-01 05:00:00         NaN


(2746033, 3)


(3126401, 3)


(3281427, 3)


(3917789, 3)


(4296067, 3)


In [106]:
taxi = pd.concat([taxi1, taxi2, taxi3, taxi4, taxi5])
print(taxi.head(3))
print('\n')
print(taxi.shape)

  Dispatching_base_num          Pickup_date  locationID
0               B00013  2015-01-01 00:30:00         NaN
1               B00013  2015-01-01 01:22:00         NaN
2               B00013  2015-01-01 01:23:00         NaN


(17367717, 3)


반복문으로 데이터 준비하기

In [110]:
import glob
nyc_taxi_data = glob.glob('../data/fhv_*')

list_taxi_df = []

for csv_filename in nyc_taxi_data:
    print(csv_filename)
    df = pd.read_csv(csv_filename)
    list_taxi_df.append(df)
    
print('\n')    
print(len(list_taxi_df))
print('\n')
print(type(list_taxi_df[0]))
print('\n')
print(list_taxi_df[0].head())
print('\n')

taxi_loop_concat = pd.concat(list_taxi_df)
print(taxi_loop_concat.shape)
print('\n')

print(taxi.equals(taxi_loop_concat))
print('\n')

../data\fhv_tripdata_2015-01.csv
../data\fhv_tripdata_2015-02.csv
../data\fhv_tripdata_2015-03.csv
../data\fhv_tripdata_2015-04.csv
../data\fhv_tripdata_2015-05.csv


5


<class 'pandas.core.frame.DataFrame'>


  Dispatching_base_num          Pickup_date  locationID
0               B00013  2015-01-01 00:30:00         NaN
1               B00013  2015-01-01 01:22:00         NaN
2               B00013  2015-01-01 01:23:00         NaN
3               B00013  2015-01-01 01:44:00         NaN
4               B00013  2015-01-01 02:00:00         NaN


(17367717, 3)


True


- 자료형 변환하기

In [2]:
import pandas as pd
import seaborn as sns

tips = sns.load_dataset("tips")
tips.iloc[:5]

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [3]:
tips['sex'].dtype

CategoricalDtype(categories=['Male', 'Female'], ordered=False)

.dtype 으로 시리즈에서의 자료형 타입 확인 가능. 위 타입은 'categories'

In [4]:
tips['sex_str'] = tips['sex'].astype(str)
print(tips.dtypes)

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


In [5]:
tips['total_bill'] = tips['total_bill'].astype(str)
print(tips.dtypes)

total_bill      object
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


In [6]:
tips['total_bill'] = tips['total_bill'].astype(float)
print(tips.dtypes)

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


- 잘못 입력한 데이터 처리하기

In [7]:
tips_sub_miss = tips.head(10)
tips_sub_miss.loc[[1,3,5,7], 'total_bill'] = 'missing'
print(tips_sub_miss)

  total_bill   tip     sex smoker  day    time  size sex_str
0      16.99  1.01  Female     No  Sun  Dinner     2  Female
1    missing  1.66    Male     No  Sun  Dinner     3    Male
2      21.01  3.50    Male     No  Sun  Dinner     3    Male
3    missing  3.31    Male     No  Sun  Dinner     2    Male
4      24.59  3.61  Female     No  Sun  Dinner     4  Female
5    missing  4.71    Male     No  Sun  Dinner     4    Male
6       8.77  2.00    Male     No  Sun  Dinner     2    Male
7    missing  3.12    Male     No  Sun  Dinner     4    Male
8      15.04  1.96    Male     No  Sun  Dinner     2    Male
9      14.78  3.23    Male     No  Sun  Dinner     2    Male


C:\Users\user\anaconda3\lib\site-packages\pandas\core\indexing.py:1720: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [8]:
print(tips_sub_miss.dtypes)

total_bill      object
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


In [9]:
# tips_sub_miss['total_bill'].astype(float) # 당연히 오류가 뜬다.

In [10]:
pd.to_numeric(tips_sub_miss['total_bill'])

ValueError: Unable to parse string "missing" at position 1

ignore은 말 그대로 오류를 무시하는 거

In [14]:
tips_sub_miss['total_bill']= pd.to_numeric(tips_sub_miss['total_bill'], errors='ignore')
print(tips_sub_miss)

  total_bill   tip     sex smoker  day    time  size sex_str
0      16.99  1.01  Female     No  Sun  Dinner     2  Female
1    missing  1.66    Male     No  Sun  Dinner     3    Male
2      21.01  3.50    Male     No  Sun  Dinner     3    Male
3    missing  3.31    Male     No  Sun  Dinner     2    Male
4      24.59  3.61  Female     No  Sun  Dinner     4  Female
5    missing  4.71    Male     No  Sun  Dinner     4    Male
6       8.77  2.00    Male     No  Sun  Dinner     2    Male
7    missing  3.12    Male     No  Sun  Dinner     4    Male
8      15.04  1.96    Male     No  Sun  Dinner     2    Male
9      14.78  3.23    Male     No  Sun  Dinner     2    Male


<ipython-input-14-c61b688ad422>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tips_sub_miss['total_bill']= pd.to_numeric(tips_sub_miss['total_bill'], errors='ignore')


In [12]:
print(tips_sub_miss.dtypes)

total_bill      object
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


erros 인자에 설정할 수 있는 값
- raise : 숫자로 변환할 수 없는 값이 있으면 오류 발생
- coerce : 숫자로 변환할 수 없는 값을 누락값으로 지정
- ignore : 아무 작업도 하지 않음

(erros 인자의 디폴트는 raise)

In [15]:
tips_sub_miss['total_bill'] = pd.to_numeric(tips_sub_miss['total_bill'], errors='coerce')
print(tips_sub_miss)

   total_bill   tip     sex smoker  day    time  size sex_str
0       16.99  1.01  Female     No  Sun  Dinner     2  Female
1         NaN  1.66    Male     No  Sun  Dinner     3    Male
2       21.01  3.50    Male     No  Sun  Dinner     3    Male
3         NaN  3.31    Male     No  Sun  Dinner     2    Male
4       24.59  3.61  Female     No  Sun  Dinner     4  Female
5         NaN  4.71    Male     No  Sun  Dinner     4    Male
6        8.77  2.00    Male     No  Sun  Dinner     2    Male
7         NaN  3.12    Male     No  Sun  Dinner     4    Male
8       15.04  1.96    Male     No  Sun  Dinner     2    Male
9       14.78  3.23    Male     No  Sun  Dinner     2    Male


<ipython-input-15-ef20acd31838>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tips_sub_miss['total_bill'] = pd.to_numeric(tips_sub_miss['total_bill'], errors='coerce')


In [16]:
print(tips_sub_miss.dtypes)

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


def to_numeric(arg, errors='raise', downcast=None)

dowoncast는 정수, 실수와 같은 자료형을 더 작은 형태로 만들때 사용한다. 이를 다운캐스트라고 한다. downcast인자에는 interger, signed, unsigned, float 등의 값을 사용할 수 있다.

In [18]:
tips_sub_miss['total_bill'] = pd.to_numeric(tips_sub_miss['total_bill'], errors='coerce', downcast='float')
print(tips_sub_miss.dtypes)

total_bill     float32
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
sex_str         object
dtype: object


<ipython-input-18-edb91e91a891>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tips_sub_miss['total_bill'] = pd.to_numeric(tips_sub_miss['total_bill'], errors='coerce', downcast='float')


float64 -> float->32 로 바뀜. 메모리 공간이 2배 줄어들음. (빅데이터를 다룰때 유용함)

- 카테고리 자료형

In [23]:
tips['sex']

0      Female
1        Male
2        Male
3        Male
4      Female
        ...  
239      Male
240    Female
241      Male
242      Male
243    Female
Name: sex, Length: 244, dtype: category
Categories (2, object): ['Female', 'Male']

In [20]:
tips['sex'] = tips['sex'].astype('str')
print(tips.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    object  
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
 7   sex_str     244 non-null    object  
dtypes: category(3), float64(2), int64(1), object(2)
memory usage: 10.8+ KB
None


In [22]:
tips['sex'] = tips['sex'].astype('category')
print(tips.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
 7   sex_str     244 non-null    object  
dtypes: category(4), float64(2), int64(1), object(1)
memory usage: 9.3+ KB
None


 10.8+ KB ->  9.3+ KB
 
 이와 같이 반복되는 문자열로 구성된 데이터는 카데고리를 사용하는 것이 더 효율적이다.
